### PR TITLE
Add the initial definitions of Size, Bound, and Vector

### PR DESCRIPTION
**Pull request by scolsen**
Vector is essentially an Array type that incorporates some amount of
bounds checking at the type-level--the bound and size types facilitate
these checks.

### CODE REVIEW

File: src/vector.carp [hunk 1]
```diff
@@ -0,0 +1,150 @@
+
+(load "https://github.com/scolsen/typeclass@master")
+(load "src/bound.carp")
+
+(doc Vector "
+  Vector is an array-like type that supports length/bounds checking at
+  type-level. It also implements the [dimensional interfaces](dimensional.html)
+  and supports generic programming.")
+(deftype (Vector s a) [elements (Array a)])
+(private Vector.elements)
```
Comment (hellerve) [inline src/vector.carp:11]: Should we make the setters private as well?

File: src/vector.carp [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        acc
+        (Vector.gen-vec (cdr vs) (cons-last (list 'Vector.push (car vs)) (cons-last '(ref) acc)))))
+
+  (doc of "
+    Creates a new Vector from the contents of an Array literal.")
+  (defmacro of [values]
+    (Vector.gen-vec values (list '==> '(Vector.empty))))
+
+  (implements positions positions)
+  (def positions )
```
Comment (hellerve) [inline src/vector.carp]: Is this valid? It looks pretty strange to me.

**Review by hellerve — APPROVED** [review]
Yay! Now Carp is on par with Agda tutorials everywhere :)

File: src/vector.carp (hunk 2 above)
Comment (scolsen) [reply src/vector.carp]: whoooops it's not--it's an accidental sneak preview of things to come

File: src/vector.carp (hunk 1 above)
Comment (scolsen) [reply src/vector.carp:11]: Good call!

**Pull request merged**